File: src/conduit/loader/csv_loader.py
"""CSV destination writer."""

from __future__ import annotations

import logging
from pathlib import Path

import pyarrow as pa
import pyarrow.csv as pcsv

from conduit.models import DestinationConfig

logger = logging.getLogger(__name__)


def load_csv(table: pa.Table, dest: DestinationConfig, base_dir: Path) -> None:
    """Write an Arrow table to a CSV file."""
    file_path = base_dir / dest.config.get("path", f"{dest.name}.csv")
    file_path.parent.mkdir(parents=True, exist_ok=True)

    write_options = pcsv.WriteOptions()
    delimiter = dest.config.get("delimiter", ",")
    if delimiter:
        write_options.delimiter = delimiter

    include_header = dest.config.get("include_header", True)
    write_options.include_header = include_header

    pcsv.write_csv(table, str(file_path), write_options=write_options)
    logger.info("Loaded %d rows to CSV destination '%s' (%s)", len(table), dest.name, file_path)

File: src/ts/loader/csv_loader.ts
/**
 * CSV destination writer.
 */

import { join, dirname } from "@std/path";
import { stringify } from "@std/csv";
import type { DataTable, DestinationConfig } from "../models.ts";

export async function loadCsv(
  table: DataTable,
  dest: DestinationConfig,
  baseDir: string,
): Promise<void> {
  const filePath = join(baseDir, (dest.config.path as string) ?? `${dest.name}.csv`);
  Deno.mkdirSync(dirname(filePath), { recursive: true });

  const delimiter = (dest.config.delimiter as string) ?? ",";
  const includeHeader = (dest.config.include_header as boolean) ?? true;

  const colNames = table.columns.map((c) => c.name);
  const rows = table.rows.map((row) => {
    const obj: Record<string, string> = {};
    for (const col of colNames) {
      const v = row[col];
      obj[col] = v === null || v === undefined ? "" : String(v);
    }
    return obj;
  });

  const csv = stringify(rows, {
    columns: includeHeader ? colNames : undefined,
    separator: delimiter,
  });

  await Deno.writeTextFile(filePath, csv);
  console.log(
    `Loaded ${table.rows.length} rows to CSV destination '${dest.name}' (${filePath})`,
  );
}
